File: Main/src/main/java/com/redelf/commons/model/HasId.java
package com.redelf.commons.model;

import androidx.annotation.Nullable;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.gson.annotations.SerializedName;

public abstract class HasId {

    @JsonProperty("id")
    @SerializedName("id")
    private Long id;

    @JsonCreator
    public HasId() {

        setId(initializeId());
    }

    @Nullable
    public Long getId() {

        return id;
    }

    public void setId(final long id) {

        this.id = id;
    }

    public boolean hasValidId() {

        return id != null && id != 0;
    }

    protected long initializeId() {

        return 0;
    }
}

File: Main/src/main/java/com/redelf/commons/model/HasId.kt
package com.redelf.commons.model

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.google.gson.annotations.SerializedName
import com.google.gson.internal.LinkedTreeMap

abstract class HasId {

    @JsonProperty("id")
    @SerializedName("id")
    private var id: Long? = null

    @JsonCreator constructor()

    @Throws(ClassCastException::class)
    constructor(data: LinkedTreeMap<String, Any>) : this() {

        id = (data["id"] as Double).toLong()
    }

    open fun getId(): Long? {

        if (id == null) {

            setId(initializeId())
        }

        return id
    }

    fun setId(id: Long) {

        this.id = id
    }

    fun hasValidId(): Boolean {

        return getId() != null && getId() != 0L
    }

    protected open fun initializeId(): Long {

        return 0
    }
}
